src/include/serial_inline.h

```diff
@@ -79,10 +81,12 @@ __insert_serial_func(WT_SESSION_IMPL *session, WT_INSERT_HEAD *ins_head, WT_INSE
          * against the next pointer might indicate that the skip list location is still valid, but
          * that may no longer be true when the atomic_cas operation executes.
          *
-         * Place a read barrier here to avoid this issue.
+         * Place a compiler barrier here to avoid this issue.
          */
-        WT_ORDERED_READ(old_ins, *ins_stack[i]);
-        if (old_ins != new_ins->next[i] || !__wt_atomic_cas_ptr(ins_stack[i], old_ins, new_ins))
+        old_ins = *ins_stack[i];
+        WT_C11_BARRIER();
+        if (old_ins != new_ins->next[i] ||
+          !__wt_atomic_memorder_cas_ptr(ins_stack[i], old_ins, new_ins, WT_ATOMIC_SEQ_CST))
             return (i == 0 ? WT_RESTART : 0);
         if (ins_head->tail[i] == NULL || ins_stack[i] == &ins_head->tail[i]->next[i])
```

Comment:
ditto all comments above. Even more so here since `ins_stack[i]` is also used after the cas. Are we intentionally forcing the compiler to reload that value? I *hope* it can't change, since I think that would have to be a data race.

Reply:
ins_stack is a thread local variable so it should be ok

Reply:
Sure but do we need to force a reload, or should it be allowed to store `ins_stack[i]` in a register? I assume we want to allow that, so we should consider assigning it to a variable like suggested in https://github.com/wiredtiger/wiredtiger/pull/9010#discussion_r1157157035

Reply:
agree